metricflow/plan_conversion/dataflow_to_sql.py

```diff
@@ -713,43 +719,38 @@ def visit_compute_metrics_node(self, node: ComputeMetricsNode) -> SqlDataSet:
 
             assert metric_expr
 
-            output_column_association = self._column_association_resolver.resolve_spec(metric_spec)
-            metric_select_columns.append(
-                SqlSelectColumn(
-                    expr=metric_expr,
-                    column_alias=output_column_association.column_name,
+            defined_from = MetricModelReference(metric_name=metric_spec.element_name)
+
+            if node.for_group_by_source_node:
+                entity_spec = output_instance_set.entity_instances[0].spec
+                group_by_metric_spec = GroupByMetricSpec(
+                    element_name=metric_spec.element_name,
+                    entity_links=(),
+                    metric_subquery_entity_links=entity_spec.entity_links + (entity_spec.reference,),
                 )
-            )
-            metric_instances.append(
-                MetricInstance(
+                output_column_association = self._column_association_resolver.resolve_spec(group_by_metric_spec)
+                group_by_metric_instance = GroupByMetricInstance(
                     associated_columns=(output_column_association,),
-                    defined_from=MetricModelReference(metric_name=metric_spec.element_name),
-                    spec=metric_spec,
+                    defined_from=defined_from,
+                    spec=group_by_metric_spec,
+                )
+            else:
+                output_column_association = self._column_association_resolver.resolve_spec(metric_spec)
+                metric_instances.append(
+                    MetricInstance(
+                        associated_columns=(output_column_association,),
+                        defined_from=defined_from,
+                        spec=metric_spec,
+                    )
                 )
+            metric_select_columns.append(
+                SqlSelectColumn(expr=metric_expr, column_alias=output_column_association.column_name)
```

Comment:
This is super awkward with the assertion on line 607 serving as a guard for single-metric-special-case logic that mostly manifests on line 751 (!!!).

I don't think we should fix it here at all - you need to merge this and move on - but this whole function has clearly gotten completely out of control.